README.md
# Screeps!
Release 2.17.12 - codename Little Rascal

AI for Screeps, version 2, completed of Grunt file and tasks for stand alone / private server play.

The project is abandoned due a complete rewrite of the entire AI which will be probably published as well at some point. However, at the current state the AI is quite capable and can get you started quite ok.

### Grunt
In order to stream files directly into the "default" folder, in game, you need to update one line in the file Gruntfile.js.

In game, on the panel Script, you should find on the bottom of it a link "Open local folder". Click it copy as a string the location of the "default" folder and paste it into Grunt, as "dest".

### Features
- Automatic spawn of all units, with the exclusion of Claimers.
- Automatic constructions of roads.
- Decent balance between units.
- Units can spread into nearby rooms to collect additional energy.
- Spawns can send Defenders and Guards remotely to support Heroes.

### Flaws
- Cannot extract minerals (I did not get there with my RCL)
- Cannot perform aggressive military operations, only defensive ones.
- Optimizes only partially Memory usage and CPU load. It's not that bad though.

### Units
- Builder, takes care of all constructions in the Spawn room.
- Claimer, to be spawned manually and manually sent to claim another room.
- Defender, perform ranged attacks and heal itself and nearby unit.
- Guard, perform close quarter attacks, quite tough unit.
- Harvester, harvests energy in the Spawn room.
- Hauler, transporte energy from Containers to Structures in the Spawn room.
- Hero, moves into nearby rooms to collect additional energy, auto-construct roads.
- Miner, substitutes Harvesters when the Spawn is capable to spawn them.
- Upgrader, takes care of upgrading the Room Controller.

### Towers
Towers are programmed to prioritize defense, otherwise they repair structures and roads in the Spawn room.

### Future improvements
- Units will be able to more efficiently use Memory.
- Room Object to keep the state of the various room.
- Flags, and a generally a more dynamic gameplay.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).
This project adheres to [Semantic Versioning](http://semver.org/).

## [2.17.12] - 2018-03-05
### Edited
- Gruntfile.js
- README.md, final words, really

## [2.17.11] - 2018-03-05
### Edited
- README.md, final words

### Changed
- config.js, Tier1 Energy Cap reduced from 1000 to 800
- prototype.creep.js, bug fix recycleAt() and improvements
- prototype.spawn.js, bug fix, added calcCreepCost() method
- role.defender.js, Defenders only perform Ranged Attack

## [2.17.10] - 2018-03-05
### Changed
- prototype.creep.js, refactored getEnergy() method
- prototype.spawn.js, new deleteExpiredCreeps() method
                      new countCreepsInRoom() method
                      renamed getAdjacentRooms() into getExits()

## [2.17.9] - 2018-03-02
### Changed
- prototype.creep.js, minimal improvements
- prototype.spawn.js, bug fix
- role.hero.js, Heros reset Workroom when they transfer all residual Energy to Structures
- role.miner.js, increased time for recycling / moving towards Spawn


## [2.17.8] - 2018-03-02
### Changed
- config.js, changed Miners' Cap to the number of Energy Sources in the room
- prototype.creep.js, refactored resetWorkroom() method
                      refactored recycleAt() method
                      refactored rechargeStructure() method
- prototype.spawn.js, refactored spawnCustomCreep() method, bug fix
                      refactored getExits() method
                      refactored hijack() method
- role.harvester.js, Harvesters do not get Energy from Containers
                     Harvesters do not move into nearby rooms
- role.hero.js, Heros do not get Energy from Containers
                improved recycling routine when in Workroom

## [2.17.7] - 2018-03-01
### Changed
- config.js, Builders cap reduced from 5 to 4
- prototype.creep.js, added resetWorkroom() method
                      refactored requestRoad() method, added param fatigue
                      refactored getEnergy() method, returns Boolean
                      refactored findStructure() method, bug fix
                      refactored rechargeStructure() method, bug fix
                      renamed details() method into getDetails()
                      removed findAndrechargeStructures() method
- role.harvester.js, refactored to enable Harvester work on nearby rooms
- role.hauler.js, refactored to share more common code with other units
- role.hero.js, refactored, bug fix and optimization

## [2.17.6] - 2018-03-01
### Changed
- config.js, better balance of units spawning
- prototype.spawn.js, refactored Spawning routine, unified into spawnCustomCreep() method
                      hardcoded nearby rooms, temporary test
- role.hero.js, bug fix

## [2.17.5] - 2018-03-01
### Changed
- prototype.creep.js, renamed recharge() method into getEnergy()
                      refactored getEnergy(), bug fix
- role.hero.js, refactored

## [2.17.4] - 2018-03-01
### Changed
- prototype.creep.js, renamed suicideAt() method into recycleAt()
                      refactored recharge() method
                      removed longRecharge() method
- role.guard.js, implemented recycledAt(30)
- role.defender.js, implemented recycleAt(30)
- role.hauler.js, implemented recharge() method
- role.hero.js, refactored under attack routine
                added recycleAt(30) method
                replaced longRecharge() method with recharge()
- role.miner.js, implemented recycleAt(20);

## [2.17.3] - 2018-02-28
### Changed
- config.js, Tier1 Energy Cap increased to 1000 Energy
- prototype.creep.js, constansts VERBOSE and DEBUG
- prototype.spawn.js, constansts VERBOSE and DEBUG
- prototype.tower.js, constansts VERBOSE and DEBUG

### Removed
- utils.js, not used any longer

## [2.17.2] - 2018-02-28
### Changed
- prototype.creep.js, refactored suicideAt() method

## [2.17.1] - 2018-02-28
### Added
- prototype.creep.js, suicideAt() method
- role.builder.js, Builder can now be recycled by Spawn when closer to end of life
- role.harvester.js, Harvesters can now be recycled by Spawn when closer to end of life
- role.hauler.js, Haulers can now be recycled by Spawn when closer to end of life
- role.upgrader.js, Upgraders can now be recycled by Spawn when closer to end of life

## [2.17.0] - 2018-02-28
### Added
- prototype.creep.js, added headForHomeroom() method
- role.guard.js, Tier3 Guard units added to the pool of available Creeps

### Changed
- config.js, includes Guards in units count
- prototype.spawn.js, refactored spawning to add Guards
- role.builder.js, Builders can now pickup dropped resources
- role.harvester.js, Harvesters can now pickup dropped resources
- role.hauler.js, Haulers no longer pickup dropped resources
- role.hero.js, Heros head for homeroom if damaged
- role.upgrader.js, Upgraders can now pickup dropped resources

## [2.16.4] - 2018-02-27
### Changed
- config.js, changes Harvesters and Hauler Cap limits.

## [2.16.3] - 2018-02-27
### Changed
- config.js, Tier 2 Energy Threshold downgraded from 1000 to 800
- prototype.creep.js, bug fix

## [2.16.2] - 2018-02-27
### Added
- role.miner.js, Unit stores energy source's and container's ids in memory

### Changed
- prototype.creep.js, refactored isLocked() method
- prototype.creep.js, rechargeStructures() method renamed into findAndRechargeStructures()

## [2.16.1] - 2018-02-25
### Added
- prototype.creep.js, new methods findStructure and rechargeStructure
- prototype.spawn.js, bug fix

### Changed
- role.harvester.js, refactored structures finding routine
- role.hero.js, refactored structures finding routine

## [2.16.0] - 2018-02-25
### Added
- role.claimer.js, first iteration for Claimer units
- prototype.spawn.js, refactored spawnTier3 method, can now spawn Claimers
- prototype.spawn.js, Spawns can now hijack Hero units on passage, resetting their homeroom and workroom

src/config.js
/**
  Creeps Tier 1
  */
const BUILDER   = "builder";
const HARVESTER = "harvester";
const UPGRADER  = "upgrader";

/**
  Creeps Tier 2
  */
const HAULER = "hauler";
const HERO   = "hero";
const MINER  = "miner";

/**
  Creeps Tier 3
  */
const CLAIMER  = "claimer";
const DEFENDER = "defender";
const GUARD    = "guard";

module.exports = {

  /**
    Global settings,
    applied to the entire Empire.
    */
  global: function() {
    return {}
  },

  /**
    Max energy available for Creeps Tier 1.
    capped to 800
    */
  tier1_energy_cap: function() {
    return 800;
  },

  /**
    Max energy available for Creeps Tier 2.
    capped to 1500
    */
  tier2_energy_cap: function() {
    return 1500;
  },

  /**
    Creeps Tier 2 allowed only at 1000 energy max capacity for current room.
    */
  tier2_energy_threshold: function() {
    return 800;
  },

  /**
    Creeps Tier 3 allowed only at 1500 energy max capacity for current room.
    */
  tier3_energy_threshold: function() {
    return 1500;
  },

  /**
    Harvesters are spawn in presence of active energy sources.
    Range [0, 8]
    @param room ROOM
    @param creepsCount Object, key/value count of creeps in the current room
    */
  harvesters_cap: function(room, creepsCount) {
    const tier2 = creepsCount[HAULER] + creepsCount[MINER] * 2;
    const sources = room.find(FIND_SOURCES_ACTIVE);
    if (sources.length > 1) return 8 - tier2;
    return 4 - tier2;
  },

  /**
    Builders are spawn only if required.
    Range [0, 5]
    */
  builders_cap: function(room) {
    // const maxEnergy = room.energyCapacityAvailable;
    const constructionSites = room.find(FIND_CONSTRUCTION_SITES);
    return _.min([constructionSites.length, 4]);
  },

  /**
    Upgraders are spawn in a max of 8,
    reduced in number if any building is required.
    Range [2, 4]
    */
  upgraders_cap: function(room) {
    const constructionSites = room.find(FIND_CONSTRUCTION_SITES);
    return _.max([2, 4 - this.builders_cap(room)]);
  },

  /**
    Haulers capped to Miners * 2
    */
  haulers_cap: function(room) {
    return this.miners_cap(room) * 2;
  },

  /**
    Miners capped to number of Energy Sources in the room
    */
  miners_cap: function(room) {
    return room.find(FIND_SOURCES).length;
  },

  /**
    Guard capped at 1 per room
    */
  guard_cap: function() {
    return 0;
  },

  /**
    Repair or maintenance on structures required
    only below a specific limit of sustained damage.
    */
  repair_threshold: function() {
    return 0.5; // 50%
  },

  /**
    Improvement, repair or maintenance on walls and ramparts
    required only up to a specific limit of hit points.
    */
  wall_threshold: function() {
    return 10000; // up 50.000 hit points
  },
  rampart_threshold: function() {
    return 10000; // up 50.000 hit points
  }
}
